// advanced-command-handler/dist/classes/Event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Event = void 0;
const CommandHandler_1 = require("../CommandHandler");
class Event {
    /**
     * @param options - Options for the event.
     * @param runFunction - The run function, executed when the event is fired.
     * {@link "@discord.js".GuildMember}
     */
    constructor(options, runFunction) {
        this.run = runFunction;
        this.name = options.name;
        this.once = options.once ?? false;
    }
    /**
     * Bind the event to the client, when the `something` event from {@link AdvancedClient} will be fire, this event will be also fired.
     *
     * @param client - The client to bind the event from.
     */
    bind(client) {
        if (this.once)
            client?.once(this.name, this.run.bind(null, CommandHandler_1.CommandHandler));
        else
            client?.on(this.name, this.run.bind(null, CommandHandler_1.CommandHandler));
    }
    /**
     * Unbinds the event to the client.
     *
     * @param client - The client to unbind the event from.
     */
    unbind(client) {
        client.removeListener(this.name, this.run.bind(null, CommandHandler_1.CommandHandler));
    }
}
exports.Event = Event;
