// advanced-command-handler/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.permissionsError = exports.dayjs = exports.argError = exports.Event = exports.CommandHandlerError = exports.CommandHandler = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
exports.dayjs = dayjs_1.default;
const CommandHandler_1 = __importDefault(require("./classes/CommandHandler"));
exports.CommandHandler = CommandHandler_1.default;
const CommandHandlerError_1 = __importDefault(require("./classes/CommandHandlerError"));
exports.CommandHandlerError = CommandHandlerError_1.default;
const Event_1 = __importDefault(require("./classes/Event"));
exports.Event = Event_1.default;
const argError_1 = __importDefault(require("./utils/argError"));
exports.argError = argError_1.default;
const permissionsError_1 = __importDefault(require("./utils/permissionsError"));
exports.permissionsError = permissionsError_1.default;
__exportStar(require("./classes/Command"), exports);
__exportStar(require("./utils/Logger"), exports);
__exportStar(require("./utils/getThing"), exports);
__exportStar(require("discord.js-better-embed"), exports);
