// advanced-command-handler/dist/utils/codeError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.codeError = void 0;
const discord_js_better_embed_1 = require("discord.js-better-embed");
const Logger_1 = require("./Logger");
const utils_1 = require("./utils");
/**
 * A function to use when a code error occurs in a command for example.
 *
 * @remarks
 * If the message author is an owner, it sends the error stack.
 *
 * @param message - The message where the error is from.
 * @param error - The native error.
 * @param command - The command to be executed.
 *
 * @returns The error message sent.
 *
 * @remarks
 * Do not use for sending string as errors.
 */
function codeError(message, error, command) {
    const embed = discord_js_better_embed_1.BetterEmbed.fromTemplate('complete', {
        client: message.client,
        color: 0xee2200,
        title: 'Code error :',
        description: utils_1.cutIfTooLong(error.stack ?? error.toString(), 2048),
    });
    Logger_1.Logger.error(error, 'CodeError');
    return utils_1.isOwner(message.author.id) ? message.channel.send(embed) : message.channel.send(`An error occurred while executing the \`${command.name}\` command.`);
}
exports.codeError = codeError;
