// advanced-command-handler/dist/utils/Logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.colors = exports.LogType = void 0;
const chalk_1 = __importDefault(require("chalk"));
const dayjs_1 = __importDefault(require("dayjs"));
const util_1 = require("util");
exports.LogType = {
    error: 'red',
    warn: 'yellow',
    info: 'blue',
    event: 'green',
    log: 'default',
    test: 'white',
    comment: 'gray',
};
exports.colors = {
    red: '#b52825',
    orange: '#e76a1f',
    gold: '#deae17',
    yellow: '#eeee23',
    green: '#3ecc2d',
    teal: '#11cc93',
    blue: '#2582ff',
    indigo: '#524cd9',
    violet: '#7d31cc',
    magenta: '#b154cf',
    pink: '#d070a0',
    brown: '#502f1e',
    black: '#000000',
    grey: '#6e6f77',
    white: '#ffffff',
    default: '#cccccc',
};
class Logger {
    /**
     * @remarks
     * Avoid using it because you can't do anything with it.
     */
    constructor() { }
    /**
     * Log a message in the console as a comment.
     *
     * @remarks
     * Using the grey color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     */
    static comment(message, title = 'comment') {
        if (Logger.logComments)
            Logger.process(message, exports.LogType.comment, title);
    }
    /**
     * Log a message in the console as an error.
     *
     * @remarks
     * Using the red color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     */
    static error(message, title = 'error') {
        Logger.process(message, exports.LogType.error, title);
    }
    /**
     * Log a message in the console as an event.
     *
     * @remarks
     * Using the green color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     */
    static event(message, title = 'event') {
        Logger.process(message, exports.LogType.event, title);
    }
    /**
     * Log a message in the console as an info.
     *
     * @remarks
     * Using the blue color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     */
    static info(message, title = 'info') {
        Logger.process(message, exports.LogType.info, title);
    }
    /**
     * Log a message in the console.
     *
     * @remarks
     * Using the # color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     * @param color - The color of the log.
     */
    static log(message, title = 'log', color = exports.LogType.log) {
        Logger.process(message, color, title);
    }
    static setColor(color = exports.colors.default, text = '') {
        let finalColor;
        if ((color = this.getColorFromColorResolvable(color)))
            finalColor = chalk_1.default.hex(color);
        else
            throw new Error('Waiting for a log type, color or HexColor but receive something else.');
        return text ? finalColor(text) : finalColor();
    }
    /**
     * Log a message in the console as a test.
     *
     * @remarks
     * Using the default color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     */
    static test(message, title = 'test') {
        Logger.process(message, exports.LogType.test, title);
    }
    /**
     * Log a message in the console as a warn.
     *
     * @remarks
     * Using the yellow color.
     *
     * @param message - The message to log, can be anything.
     * @param title - The title of the log.
     */
    static warn(message, title = 'warn') {
        Logger.process(message, exports.LogType.warn, title);
    }
    /**
     * Log something in the console and transform the ColorResolvable into a ASCII Escape Sequence containing the color.
     *
     * @param text - The text to log.
     * @param color - The color of the text.
     * @param title - The title of the text.
     * @internal
     */
    static process(text, color = 'test', title = '') {
        text = typeof text === 'string' ? text : util_1.inspect(text);
        text = text.replace(/(?<![;\d])\d+(\.\d+)?(?!;|\d)/g, (match) => chalk_1.default.yellow(match));
        text = text.replace(/\u001b\[\u001b\[33m39\u001b\[39mm/gi, chalk_1.default.reset());
        color = Logger.propertyInEnum(exports.LogType, color) ?? color;
        text = `${Logger.setColor('#847270', `[${dayjs_1.default().format('D HH:mm:ss.SSS')}]`)}${Logger.setColor(color, `[${title.toUpperCase()}] ${text + chalk_1.default.reset()}`)}`;
        console.log(text);
    }
    /**
     * Returns a color in hexadecimal without the sharp from a ColorResolvable.
     *
     * @remarks
     * Returns the default color if it cannot be resolved.
     *
     * @param color - The ColorResolvable.
     * @returns The color.
     * @internal
     */
    static getColorFromColorResolvable(color) {
        return (Logger.propertyInEnum(exports.LogType, Logger.propertyInEnum(exports.colors, color) ?? '') ??
            Logger.propertyInEnum(exports.colors, color) ??
            Logger.propertyInEnum(exports.LogType, color)?.match(/#[0-9|a-f]{6}/i)?.[0] ??
            color.match(/#[0-9|a-f]{6}/i)?.[0] ??
            exports.colors.default.substring(1, 7));
    }
    /**
     * Get the value of an enum.
     *
     * @typeParam V - An object.
     * @param enumObject - The enum as an object.
     * @param property - The property to get.
     * @returns The value from the key of the enum or undefined if not found.
     * @internal
     */
    static propertyInEnum(enumObject, property) {
        return enumObject[property] ?? undefined;
    }
}
exports.Logger = Logger;
/**
 * If this property is set to `true` the {@link Logger.comment} method won't send logs.
 */
Logger.logComments = true;
