// advanced-command-handler/dist/utils/permissionsError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.permissionsError = void 0;
const discord_js_better_embed_1 = require("discord.js-better-embed");
/**
 * A function to use when an user or the client hasn't all the permissions needed.
 *
 * @param message - The message where the error is from.
 * @param missingPermissions - The error.
 * @param command - The command to be executed.
 * @param fromClient - If the error is from the client.
 *
 * @returns The error message sent.
 */
function permissionsError(message, missingPermissions, command, fromClient = false) {
    const embed = discord_js_better_embed_1.BetterEmbed.fromTemplate('complete', {
        client: message.client,
        color: 0xee2200,
        title: 'Permissions error :',
        description: `${fromClient ? 'The bot is' : 'You are'} missing permissions :\n\`${missingPermissions.sort().join('`\n`')}\``,
    });
    if (command.usage)
        embed.addField('Syntax :', command.usage);
    return message.channel.send(embed);
}
exports.permissionsError = permissionsError;
